src/frontend/screens/Settings/MapSettings/MapImportBottomSheet.tsx

```diff
@@ -0,0 +1,322 @@
+          {"( .mbtiles )"}
+        </Text>
+      </TouchableOpacity>
+    </BottomSheetContent>
+  );
+};
+
+/**
+ * Bottom sheet content for when an imported file is processing, checking the
+ * format etc. is correct (shows until the import endpoint returns with the
+ * import id). User is unable to close this sheet.
+ */
+const ImportingContent = () => {
+  return <BottomSheetContent title="Processing file..." buttonConfigs={[]} />;
+};
+
+/**
+ * Bottom sheet content for when there is an error with the selected file
+ */
+const FileErrorContent = ({ onPressClose }: { onPressClose: () => void }) => {
+  const { formatMessage: t } = useIntl();
+  return (
+    <BottomSheetContent
+      title={t(m.importErrorTitle)}
+      icon={<ErrorIcon color={RED} size={90} style={styles.errorIcon} />}
+      description={t(m.fileErrorDescription)}
+      buttonConfigs={[
+        {
+          onPress: onPressClose,
+          text: t(m.close),
+          variation: "outlined",
+        },
+      ]}
+    />
+  );
+};
+
+type BottomSheetStates = "select" | "importing" | "error";
+
+export interface MapImportBottomSheetMethods {
+  /** Open the bottom sheet by expanding it to its maximum snap point */
+  open: () => void;
+}
+
+/** Custom backdrop that will close sheet when pressed */
+const BackdropPressable = (props: BottomSheetBackdropProps) => {
+  return (
+    <BottomSheetBackdrop {...props} opacity={0.3} pressBehavior={"close"} />
+  );
+};
+
+/** Custom backdrop that will not close sheet when pressed */
+const BackdropNonPressable = (props: BottomSheetBackdropProps) => {
+  return (
+    <BottomSheetBackdrop {...props} opacity={0.3} pressBehavior={"none"} />
+  );
+};
+
+const MapImportBottomSheet = React.forwardRef<MapImportBottomSheetMethods, {}>(
+  (_props, ref) => {
+    const [state, setState] = React.useState<BottomSheetStates>("select");
+    const bottomSheetRef = React.useRef<BottomSheetMethods>(null);
+    const { snapPoints, updateSheetHeight } = useSnapPointsCalculator();
+    const { add: addMapImport } = useMapImportsManager();
+
+    React.useImperativeHandle(ref, () => {
+      return {
+        open: () => bottomSheetRef.current?.expand(),
+      };
+    });
+
+    const closeSheet = () => {
+      // Set state on close so when it opens again this is what you see
+      setState("select");
+      bottomSheetRef.current?.close();
+    };
+
+    const handlePressImport = async () => {
+      setState("importing");
+      let results;
+      try {
+        results = await DocumentPicker.getDocumentAsync();
+      } catch (err) {
+        setState("error");
+        return;
+      }
+
+      if (results.type === "cancel") {
+        closeSheet();
+        return;
+      }
+
+      try {
+        const {
+          import: { id: importId },
+          style,
+        } = await api.maps.importTileset(results.uri);
+
+        if (style) {
+          addMapImport({ styleId: style.id, importId });
+        }
```

Comment:
Is there a case where there would be no style?

Reply:
in practice, not currently. the TS typing has style as potentially null though so this is there to satisfy that. otherwise, i believe the imports we're dealing with will always generate a style